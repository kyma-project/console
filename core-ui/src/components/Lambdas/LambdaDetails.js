import React from 'react';

import LambdaDetailsWrapper from 'components/Lambdas/LambdaDetails/LambdaDetailsWrapper';
import { useConfigData } from 'components/Lambdas/helpers/misc/useConfigData';

export default function LambdaDetails({ match }) {
  useConfigData();
  return <LambdaDetailsWrapper lambdaName={match.params.name} />;
}
