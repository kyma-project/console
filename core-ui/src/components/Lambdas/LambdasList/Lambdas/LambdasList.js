import React from 'react';
import LuigiClient from '@luigi-project/client';

import { GenericList, Labels } from 'react-shared';
import { Link } from 'fundamental-react';
import { LambdaStatusBadge } from 'components/Lambdas/LambdaStatusBadge/LambdaStatusBadge';

import CreateLambdaModal from './CreateLambdaModal';
import { useDeleteLambda } from 'components/Lambdas/gql/hooks/mutations';

import { prettySourceType } from 'components/Lambdas/helpers/lambdas';
import { prettyRuntime } from 'components/Lambdas/helpers/runtime';
import {
  SERVERLESS_PRESETS_LABELS,
  LAMBDAS_LIST,
  ERRORS,
  REFETCH_LAMBDAS_TIMEOUT,
} from 'components/Lambdas/constants';

const headerRenderer = () => [
  'Name',
  'Source Type',
  'Runtime',
  'Labels',
  'Status',
];
const textSearchProperties = [
  'name',
  'prettySourceType',
  'prettyRuntime',
  'labels',
  'status.phase',
];

export default function LambdasList({
  lambdas = [],
  repositories,
  serverDataError = false,
  serverDataLoading = false,
  refetchLambdas,
}) {
  const functionNames = lambdas.map(l => l.name);
  const headerActions = (
    <CreateLambdaModal
      functionNames={functionNames}
      repositories={repositories}
      serverDataError={serverDataError || false}
      serverDataLoading={serverDataLoading || false}
    />
  );

  const deleteLambda = useDeleteLambda({
    redirect: false,
    onSuccessCallback: () => {
      setTimeout(() => {
        refetchLambdas();
      }, REFETCH_LAMBDAS_TIMEOUT);
    },
  });

  const actions = [
    {
      name: 'Delete',
      handler: lambda => {
        deleteLambda(lambda);
      },
    },
  ];
  const rowRenderer = lambda => [
    <Link
      className="link"
      data-test-id="lambda-name"
      onClick={() =>
        LuigiClient.linkManager().navigate(`details/${lambda.name}`)
      }
    >
      {lambda.name}
    </Link>,
    <span>{prettySourceType(lambda.sourceType)}</span>,
    <span>{prettyRuntime(lambda.runtime)}</span>,
    <Labels labels={lambda.labels} ignoreLabels={SERVERLESS_PRESETS_LABELS} />,
    <LambdaStatusBadge status={lambda.status} />,
  ];

  return (
    <GenericList
      actions={actions}
      entries={lambdas.map(lambda => ({
        ...lambda,
        prettyRuntime: prettyRuntime(lambda.runtime),
        prettySourceType: prettySourceType(lambda.sourceType),
      }))}
      showSearchField={true}
      showSearchSuggestion={false}
      textSearchProperties={textSearchProperties}
      extraHeaderContent={headerActions}
      headerRenderer={headerRenderer}
      rowRenderer={rowRenderer}
      serverDataError={serverDataError}
      serverDataLoading={serverDataLoading}
      notFoundMessage={LAMBDAS_LIST.ERRORS.RESOURCES_NOT_FOUND}
      noSearchResultMessage={LAMBDAS_LIST.ERRORS.NOT_MATCHING_SEARCH_QUERY}
      serverErrorMessage={ERRORS.SERVER}
    />
  );
}
