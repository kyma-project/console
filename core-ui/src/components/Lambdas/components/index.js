export * from './LabelsInput/LabelsInput';
export * from './LambdaNameInput/LambdaNameInput';
export * from './TabsWithActions/TabsWithActions';
