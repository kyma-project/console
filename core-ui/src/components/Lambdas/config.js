const defaultConfig = {
  functionUsageKind: 'serverless-function',
};

function getConfigValue(field) {
  const serverlessConfig = window.clusterConfig?.serverless;
  const defaultValue = defaultConfig[field] || '';

  if (!serverlessConfig) {
    return defaultValue;
  }

  return serverlessConfig[field] || defaultValue;
}

function loadConfig() {
  return {
    functionUsageKind: getConfigValue('functionUsageKind'),
  };
}

export const CONFIG = loadConfig();
