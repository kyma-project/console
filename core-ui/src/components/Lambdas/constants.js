export const BUTTONS = {
  CANCEL: 'Cancel',
  DELETE: 'Delete',
};

export const ERRORS = {
  SERVER: 'Server error. Contact your cluster admin.',
  RESOURCES_NOT_FOUND: "This lambda doesn't have any related resources yet.",
  NOT_MATCHING_SEARCH_QUERY: "Couldn't find resources matching this query.",
  NOT_MATCHING_FILTERS: "Couldn't find resources matching these filters.",
};

export const TOOLBAR = {
  TITLE: 'Lambdas',
  DESCRIPTION:
    'Extend your applications with lambdas that you can trigger with incoming events and expose outside the cluster with API Rules.',
};

export const TRIGGER_SUBSCRIBER = {
  SERVING_SERVICE: 'Service',
  SERVING_API_VERSION: 'serving.knative.dev/v1',
  BROKER: 'default',
};

export const GQL_QUERIES = {
  EVENT_ACTIVATIONS: {
    ERROR_MESSAGE: `Error while fetching available Events in "{namespace}" namespace: {error}`,
  },
  EVENT_TRIGGERS: {
    ERROR_MESSAGE: `Error while fetching Event Triggers for "{lambdaName}" Lambda: {error}`,
  },
  SERVICE_INSTANCES: {
    ERROR_MESSAGE: `Error while fetching Service Instances: {error}`,
  },
};

export const GQL_MUTATIONS = {
  CREATE_TRIGGERS: {
    SUCCESS_MESSAGE_SINGLE: `Event Trigger created successfully`,
    SUCCESS_MESSAGE_MANY: `Event Triggers created successfully`,
    ERROR_MESSAGE_SINGLE: `Error while creating an Event Trigger for "{lambdaName}" Lambda: {error}`,
    ERROR_MESSAGE_MANY: `Error while creating Event Triggers for "{lambdaName}" Lambda: {error}`,
  },
  DELETE_TRIGGER: {
    SUCCESS_MESSAGE: `Event Trigger was successfully removed`,
    ERROR_MESSAGE: `Error while deleting "{triggerName}" Event Trigger for "{lambdaName}" Lambda: {error}`,
    CONFIRM_MODAL: {
      TITLE: `Remove Event Trigger`,
      MESSAGE: `Are you sure you want to delete "{triggerName}" Event Trigger for "{lambdaName}" Lambda?`,
    },
  },
};

export const TRIGGER_SCHEMA = {
  CSS_PREFIX: 'schema-event-trigger',
  PAYLOAD_TEXT: 'Payload',
  EXAMPLE_TEXT: 'Example',
  COLUMN_NAMES: ['Name', 'Title', 'Type', 'Format', 'Default', 'Description'],
};

export const EVENT_TRIGGERS_PANEL = {
  LIST: {
    TITLE: 'Event Triggers',
    ERRORS: {
      RESOURCES_NOT_FOUND:
        "This lambda doesn't have any Event Triggers yet.",
      NOT_MATCHING_SEARCH_QUERY:
        "Couldn't find Event Triggers matching this query.",
    },
  },
  ADD_MODAL: {
    OPEN_BUTTON: {
      TEXT: 'Add Event Trigger',
      NOT_ENTRIES_POPUP_MESSAGE:
        'No Events available to connect in current namespace.',
    },
    TITLE: 'Add Event Trigger',
    CONFIRM_BUTTON: {
      TEXT: 'Add',
      INVALID_POPUP_MESSAGE: 'At least one Event must be checked.',
    },
  },
};

export const SERVICE_BINDINGS_PANEL = {
  LIST: {
    TITLE: 'Service Bindings',
    ERRORS: {
      RESOURCES_NOT_FOUND:
        "This lambda doesn't have any Service Bindings yet.",
      NOT_MATCHING_SEARCH_QUERY:
        "Couldn't find Service Bindings matching this query.",
    },
  },
  CREATE_MODAL: {
    OPEN_BUTTON: {
      TEXT: 'Create Service Binding',
      NOT_ENTRIES_POPUP_MESSAGE:
        'There are no Service Instances available to bind in this Namespace. Create a Service Instance first.',
    },
    TITLE: 'Create Service Binding',
    CONFIRM_BUTTON: {
      TEXT: 'Create',
      POPUP_MESSAGES: {
        NO_SERVICE_INSTANCE_SELECTED: 'You must select a Service Instance.',
        NO_SECRET_SELECTED: 'You must create or select a Secret.',
      },
    },
  },
  FORM: {
    NO_SECRETS_FOUND:
      'There are no Secrets available. Create a new Secret to bind the Service Instance.',
  },
};
