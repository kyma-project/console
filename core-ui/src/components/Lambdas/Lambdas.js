import React from 'react';
import ModalWithForm from '../ModalWithForm/ModalWithForm';
import CreateLambdaForm from './CreateLambdaForm/CreateLambdaForm';
import { useQuery, useMutation } from '@apollo/react-hooks';
import LuigiClient from '@kyma-project/luigi-client';
import { GET_LAMBDAS } from '../../gql/queries';
import {
  DELETE_LAMBDA,
  DELETE_SERVICE_BINDING_USAGES,
} from '../../gql/mutations';

import { GenericList, useNotification } from 'react-shared';

import extractGraphQlErrors from 'shared/graphqlErrorExtractor';

import builder from '../../commons/builder';
import { Spinner, Labels } from 'react-shared';
import LambdaStatusBadge from '../../shared/components/LambdaStatusBadge/LambdaStatusBadge';
import { REFETCH_TIMEOUT } from '../../shared/constants';
import { PageHeader } from 'react-shared';

function CreateLambdaModal() {
  return (
    <ModalWithForm
      title="Create new lambda"
      button={{ text: 'Create lambda', option: 'light' }}
      id="add-lambda-modal"
      renderForm={props => <CreateLambdaForm {...props} />}
    />
  );
}

export default function Lambdas() {
  const { data, error, loading, refetch } = useQuery(GET_LAMBDAS, {
    variables: {
      namespace: builder.getCurrentEnvironmentId(),
    },
    fetchPolicy: 'no-cache',
  });

  const [deleteLambdaServiceBindingUsages] = useMutation(
    DELETE_SERVICE_BINDING_USAGES,
  );
  const [deleteLambda] = useMutation(DELETE_LAMBDA);
  const notificationManager = useNotification();

  if (error) {
    return `Error! ${error.message}`;
  }

  if (loading) {
    return <Spinner />;
  }

  function handleError(error, name) {
    const errorToDisplay = extractGraphQlErrors(error);
    notificationManager.notifyError({
      content: `Error while removing lambda ${name}: ${errorToDisplay}`,
      autoClose: false,
    });
  }

  const handleLambdaDelete = (lambda, refetch) => {
    const name = lambda.name;
    const namespace = lambda.namespace;

    LuigiClient.uxManager()
      .showConfirmationModal({
        header: `Remove ${name}`,
        body: `Are you sure you want to delete lambda "${name}"?`,
        buttonConfirm: 'Delete',
        buttonDismiss: 'Cancel',
      })
      .then(async () => {
        try {
          const deletedLambda = await deleteLambda({
            variables: {
              name,
              namespace,
            },
          });

          if (deletedLambda.error) {
            handleError(deletedLambda.error, name);
            return;
          }

          const serviceBindingUsageNames = (
            lambda.serviceBindingUsages || []
          ).map(s => s.name);
          if (serviceBindingUsageNames.length) {
            const deletedBindingUsages = await deleteLambdaServiceBindingUsages(
              {
                variables: { serviceBindingUsageNames, namespace },
              },
            );

            if (deletedBindingUsages.error) {
              handleError(deletedLambda.error);
              return;
            }
          }

          notificationManager.notifySuccess({
            content: `Lambda ${name} deleted`,
            autoClose: true,
          });
          setTimeout(() => {
            refetch();
          }, REFETCH_TIMEOUT);
        } catch (e) {
          handleError(e, name);
        }
      })
      .catch(e => {
        notificationManager.notifyError({
          content: `Problem with Luigi: ${e.message}`,
          autoClose: false,
        });
      });
  };

  const actions = [
    {
      name: 'Delete',
      handler: entry => {
        handleLambdaDelete(entry, refetch);
      },
    },
  ];

  const headerRenderer = () => ['Name', 'Runtime', 'Labels', 'Status'];

  const rowRenderer = item => [
    <span
      className="link"
      data-test-id="lambda-name"
      onClick={() => LuigiClient.linkManager().navigate(`details/${item.name}`)}
    >
      {item.name}
    </span>,
    <span>{item.runtime}</span>,
    <Labels labels={item.labels} />,
    <LambdaStatusBadge status={item.status} />,
  ];

  return (
    <>
      <PageHeader title="Lambdas" />
      <GenericList
        actions={actions}
        entries={data.functions}
        headerRenderer={headerRenderer}
        rowRenderer={rowRenderer}
        extraHeaderContent={<CreateLambdaModal />}
      />
    </>
  );
}
