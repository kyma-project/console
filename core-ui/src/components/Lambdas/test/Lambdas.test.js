import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import {
  allLambdasQuery,
  deleteLambdaMutation,
} from '../../../testing/queriesMocks';
import { MockedProvider } from '@apollo/react-testing';
import { componentUpdate } from '../../../testing';
import { act } from 'react-dom/test-utils';

import Lambdas from './../Lambdas';
import { Spinner } from 'react-shared';

jest.mock('@kyma-project/luigi-client', () => ({
  uxManager: () => ({
    showConfirmationModal: () => Promise.resolve(),
    addBackdrop: () => {},
    removeBackdrop: () => {},
  }),
}));

jest.mock(
  'popper.js',
  () =>
    class Popper {
      static placements = ['left'];

      constructor() {
        return {
          destroy: () => {},
          scheduleUpdate: () => {},
        };
      }
    },
);

describe('Lambdas', () => {
  it('Renders with error in no query is mocked', async () => {
    let component;
    await renderer.act(async () => {
      component = renderer.create(
        <MockedProvider>
          <Lambdas />
        </MockedProvider>,
      );
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shows loading indicator only when data is not yet loaded', async () => {
    const component = mount(
      <MockedProvider>
        <Lambdas />
      </MockedProvider>,
    );
    expect(component.find(Spinner)).toHaveLength(1);

    await componentUpdate(component);

    expect(component.find(Spinner)).toHaveLength(0);
  });

  it('Displays lambdas with their corresponding names in the table', async () => {
    const component = mount(
      <MockedProvider mocks={[allLambdasQuery]}>
        <Lambdas />
      </MockedProvider>,
    );

    await componentUpdate(component);

    const displayedLambdaNames = component.find('[data-test-id="lambda-name"]');

    expect(displayedLambdaNames).toHaveLength(2);
    expect(displayedLambdaNames.at(0).exists()).toBe(true);
    expect(displayedLambdaNames.at(1).exists()).toBe(true);
    expect(displayedLambdaNames.at(0).text()).toEqual('demo');
    expect(displayedLambdaNames.at(1).text()).toEqual('demo2');
  });

  it('Validate if modal delete button fires deleteMutation', async () => {
    const component = mount(
      <MockedProvider mocks={[allLambdasQuery, deleteLambdaMutation]}>
        <Lambdas />
      </MockedProvider>,
    );

    await componentUpdate(component);

    const deleteButton = component
      .find('.fd-button--light.sap-icon--delete')
      .first();

    await act(async () => {
      await deleteButton.simulate('click');
    });

    await componentUpdate(component);
    expect(deleteLambdaMutation.result).toHaveBeenCalled();
  });
});
