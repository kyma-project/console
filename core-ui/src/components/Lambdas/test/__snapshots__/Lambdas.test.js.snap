// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambdas Renders with error in no query is mocked 1`] = `
"Error! Network error: No more mocked responses for the query: query Functions($namespace: String!) {
  functions(namespace: $namespace) {
    name
    namespace
    labels
    runtime
    size
    status
    serviceBindingUsages {
      name
      __typename
    }
    __typename
  }
}
, variables: {\\"namespace\\":\\"production\\"}"
`;
